### PR TITLE
 Resource module for NX-OS static routes (nxos_static_routes)

### PR DESCRIPTION
**Pull request by adharshsrivatsr**
Depends-On: https://github.com/ansible/ansible-zuul-jobs/pull/262/

##### SUMMARY
Added nxos_static_routes resource module

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
modules/network/nxos/nxos_static_routes.py

### CODE REVIEW

**Comment by NilashishC** [comment]
recheck

**Review by NilashishC — CHANGES_REQUESTED** [review]
- "rtt.yaml" test is missing
- IMO, it would be a good idea to have multiple entries in each level in `config` key. For e.g., at least 2 AFI/SAFI entries in each vrf/global entry, at least 2 destinations in each `routes`, at least 2 exit points in each `next_hops`.
- All IP addresses should be to documentation IP range in https://tools.ietf.org/html/rfc5737 and https://tools.ietf.org/html/rfc3849, you can use something like https://www.site24x7.com/tools/ipv4-subnetcalculator.html for subnetting the allocated blocks.

- Please add tests with IPv6 destinations.

**Comment by mikewiebe** [comment]
Are the integration tests functional yet?  I just tried to run these tests against my N9k and I see the following errors:

```
TASK [nxos_static_routes : run test cases (connection=network_cli)] *******************************************************************************************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_static_routes/tasks/cli.yaml:16
included: /root/agents-ci/ansible/test/integration/targets/nxos_static_routes/tests/cli/merged.yml for n9k
fatal: [n9k]: FAILED! => {
    "reason": "Syntax Error while loading YAML.\n  mapping values are not allowed in this context\n\nThe error appears to be in '/root/agents-ci/ansible/test/integration/targets/nxos_static_routes/tests/cli/rendered.yml': line 6, column 27, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n    - name: Rendered\n        nxos_static_routes: & rendered\n                          ^ here\n"
}
fatal: [n9k]: FAILED! => {
    "reason": "Syntax Error while loading YAML.\n  mapping values are not allowed in this context\n\nThe error appears to be in '/root/agents-ci/ansible/test/integration/targets/nxos_static_routes/tests/cli/gathered.yml': line 9, column 23, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n        - name: Gather static_routes facts\n            nxos_facts: & facts\n                      ^ here\n"
}
included: /root/agents-ci/ansible/test/integration/targets/nxos_static_routes/tests/cli/deleted.yml for n9k
fatal: [n9k]: FAILED! => {
    "reason": "conflicting action statements: assert, always\n\nThe error appears to be in '/root/agents-ci/ansible/test/integration/targets/nxos_static_routes/tests/cli/rtt.yml': line 80, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - name: Assert that config was reverted\n      ^ here\n"
}
fatal: [n9k]: FAILED! => {
    "reason": "Syntax Error while loading YAML.\n  mapping values are not allowed in this context\n\nThe error appears to be in '/root/agents-ci/ansible/test/integration/targets/nxos_static_routes/tests/cli/parsed.yml': line 9, column 23, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n        - name: Gather static_routes facts\n            nxos_facts: &facts\n                      ^ here\n"
}
fatal: [n9k]: FAILED! => {
    "reason": "Syntax Error while loading YAML.\n  mapping values are not allowed in this context\n\nThe error appears to be in '/root/agents-ci/ansible/test/integration/targets/nxos_static_routes/tests/cli/overridden.yml': line 22, column 35, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n                    routes:\n                      - dest: 2001: db8: 12: : / 32\n                                  ^ here\n"
}
fatal: [n9k]: FAILED! => {
    "reason": "Syntax Error while loading YAML.\n  mapping values are not allowed in this context\n\nThe error appears to be in '/root/agents-ci/ansible/test/integration/targets/nxos_static_routes/tests/cli/replaced.yml': line 9, column 22, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n        - name: Gather static_routes post facts\n           nxos_facts:\n                     ^ here\n"
}

PLAY RECAP ****************************************************************************************************************************************************************************************************************************************************
n9k        : ok=29   changed=0    unreachable=0    failed=6    skipped=11   rescued=0    ignored=0   
```

Here is my playbook and other settings:

```yaml
---
- hosts: nxos
  gather_facts: no
  connection: local

  vars:
    debug: false

  roles:
    - { role: nxos_static_routes }
```

N9K is running version: `9.2(2)`

**Review by mikewiebe — CHANGES_REQUESTED** [review]
I did some manual testing of the PR on my n9k and see the following issues.

```
Nexus 9000v is a demo version of the Nexus Operating System

Software
  BIOS: version 
  NXOS: version 7.0(3)I7(3)
  BIOS compile time:  
  NXOS image file is: bootflash:///nxos.7.0.3.I7.3.bin
  NXOS compile time:  2/12/2018 13:00:00 [02/12/2018 19:13:48]
```

* Test 1 - State Merged

```yaml
    - name: Merge new static route configuration
      nxos_static_routes:
        config:
          - vrf: trial_vrf
            address_families:
              - afi: ipv4
                routes:
                  - dest: 192.0.2.64/24
                    next_hops:
                      - forward_router_address: 192.0.2.22
                        tag: 4
                        admin_distance: 2
```
The route gets configured but it's not idempotent.

```
n9k# show run | section 'vrf context trial_vrf'
vrf context trial_vrf
  ip route 192.0.2.0/24 192.0.2.22 tag 4 2
n9k# 
```

When I run the playbook again I see this:

```
    "changed": true, 
    "commands": [
        "vrf context trial_vrf", 
        "ip route 192.0.2.64/24 192.0.2.22 tag 4 2"
    ], 
```

* Test 2 - State Deleted

Starting state:

```
vrf context trial_vrf
  ip route 192.0.2.0/24 192.0.2.22 tag 4 2
```

Playbook:

```yaml
    - name: Merge new static route configuration
      nxos_static_routes:
        state: deleted
        config:
          - vrf: trial_vrf
            address_families:
              - afi: ipv4
                routes:
                  - dest: 192.0.2.64/24
                    next_hops:
                      - forward_router_address: 192.0.2.22
                        tag: 4
                        admin_distance: 2
```

Playbook result:

```
ok: [n9k.cisco.com] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    }, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "config": [
                {
                    "address_families": [
                        {
                            "afi": "ipv4", 
                            "routes": [
                                {
                                    "dest": "192.0.2.64/24", 
                                    "next_hops": [
                                        {
                                            "admin_distance": 2, 
                                            "dest_vrf": null, 
                                            "forward_router_address": "192.0.2.22", 
                                            "interface": null, 
                                            "route_name": null, 
                                            "tag": 4, 
                                            "track": null
                                        }
                                    ]
                                }
                            ]
                        }
                    ], 
                    "vrf": "trial_vrf"
                }
            ], 
            "running_config": null, 
            "state": "deleted"
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ****************************************************************************************************************************************************************************************************************************************************
n9k.cisco.com  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

Route is *not* deleted.

```
vrf context trial_vrf
  ip route 192.0.2.0/24 192.0.2.22 tag 4 2
```

File: test/integration/targets/nxos_static_routes/tests/cli/deleted.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  - block:
+      - name: Gather static_routes facts
+        nxos_facts: &facts
+            gather_subset:
+                - '!all'
+                - '!min'
+            gather_network_resources: static_routes
+
+      - name: Deleted
+        nxos_static_routes: &deleted
```
Comment (mikewiebe) [inline test/integration/targets/nxos_static_routes/tests/cli/deleted.yml]: Is it possible to delete a single route?  If so, please add a test for that.

File: test/integration/targets/nxos_static_routes/tests/cli/merged.yml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                            route_name: new_route
+                            
+                      - dest: 192.0.2.80/28
+                        next_hops:
+                        - forward_router_address: 192.0.2.26
+                          tag: 12
+            state: merged
+          register: result
+
+        - assert:
```
Comment (mikewiebe) [inline test/integration/targets/nxos_static_routes/tests/cli/merged.yml]: Can  you also add a test where you update an existing route?  What happens when you try and change a few parameters for a route that already exists?

File: test/integration/targets/nxos_static_routes/tests/cli/merged.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
+---
+    - debug:
+        msg: "Start nxos_static_routes merged integration tests connection={{ansible_connection}}"
+
+    - include_tasks: populate_config.yaml
+
+    - include_tasks: remove_config.yaml
```
Comment (mikewiebe) [inline test/integration/targets/nxos_static_routes/tests/cli/merged.yml]: Why are you populating and then removing the config?

File: test/integration/targets/nxos_static_routes/tests/cli/overridden.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  - block:
+      - name: Gather static_routes facts
+        nxos_facts: &facts
+          gather_subset:
+            - '!all'
+            - '!min'
+          gather_network_resources: static_routes
+
+      - name: Overridden
```
Comment (mikewiebe) [inline test/integration/targets/nxos_static_routes/tests/cli/overridden.yml]: Again here, please add a test to update an existing route property or multiple properties with state `overridden` to make sure it gets processed properly.

**Review by mikewiebe — CHANGES_REQUESTED** [review]

**Comment by adharshsrivatsr** [comment]
The integration tests are not functional yet as we faced issues with the ipv6 routes.

**Comment by mikewiebe** [comment]
Just took this PR out for another test drive and it's looking better but noticed a few behaviors that need to be clarified.

1. State `overridden` deletes default routes under the `management` vrf which isolates the testbed from the network.  IMO, similar to how we protect the default vrf for `nxos_vlans` we should discuss protection of a default route `ip route 0.0.0.0/0` under the `management` vrf.

2. With these newer resource modules it's difficult to determine what the significant keys are when specifying routes to delete.  We need a way to clearly identify resource keys vs parameters for deleting routes in the documentation.

For example:

```yaml
          - address_families:
              - afi: ipv4
                routes:
                  - dest: 192.0.2.16/28
                    next_hops:
                      - forward_router_address: 192.0.2.24
                        route_name: new_route

                  - dest: 192.0.2.80/28
                    next_hops:
                    - forward_router_address: 192.0.2.26
                      tag: 12
```

It seems to be true that `dest` is the only key that matters for deleting routes.  With the playbook entry above when state is deleted the following routes would be deleted by the task above.

```
ip route 192.0.2.80/28 192.0.2.27 tag 55
```

**Comment by adharshsrivatsr** [comment]
> @NilashishC @adharshsrivatsr
> Just took this PR out for another test drive and it's looking better but noticed a few behaviors that need to be clarified.
> 
> 1. State `overridden` deletes default routes under the `management` vrf which isolates the testbed from the network.  IMO, similar to how we protect the default vrf for `nxos_vlans` we should discuss protection of a default route `ip route 0.0.0.0/0` under the `management` vrf.
> 2. With these newer resource modules it's difficult to determine what the significant keys are when specifying routes to delete.  We need a way to clearly identify resource keys vs parameters for deleting routes in the documentation.
> 
> For example:
> 
> ```yaml
>           - address_families:
>               - afi: ipv4
>                 routes:
>                   - dest: 192.0.2.16/28
>                     next_hops:
>                       - forward_router_address: 192.0.2.24
>                         route_name: new_route
> 
>                   - dest: 192.0.2.80/28
>                     next_hops:
>                     - forward_router_address: 192.0.2.26
>                       tag: 12
> ```
> 
> It seems to be true that `dest` is the only key that matters for deleting routes. With the playbook entry above when state is deleted the following routes would be deleted by the task above.
> 
> ```
> ip route 192.0.2.80/28 192.0.2.27 tag 55
> ```

We are discussing about the management vrf issue as you said. Will update you soon.

**Comment by adharshsrivatsr** [comment]
Yes, the management vrf needs to be isolated. We are discussing about the isolation issue. Will update you soon.

For the deleted state, we have made it more granular. The user can specify either to delete the whole vrf (or) vrf + afi only (or) vrf + afi + dest (or)  vrf + afi + dest + next_hops. Please refer [this](https://github.com/ansible/ansible/blob/5fdf4063ac10a1817ced79e1593ac552cf2934d9/test/integration/targets/nxos_static_routes/tests/cli/deleted.yml). Please let me know if there is anything else.

**Comment by justjais** [comment]
Closing the PR as the content is moved to respective collection

**Pull request closed without merging**